docs/proposals/20200224-apply.md

```diff
@@ -0,0 +1,205 @@
+
+* is the cluster provisioned
+* is a specific node provisioned, ready, and responsive
+* is the actual Kubernetes version matching the desired version
+* are the critical control plane pods running and healthy
+* is the `etcd` ring healthy and in the expected state (all control plane nodes joined)
+* is the actual cluster configuration matching the expected cluster configuration
+* are worker nodes running and responsive
+
+Some of the mentioned checks are already implemented in some form in `upgrade` and
+`status` subcommands.
+
+### Reconciling The Cluster
+
+Depending on the actual state of the cluster determined by probes, KubeOne will try
+to modify the cluster to match the expected state.
+
+Some actions will be taken automatically, however, it is not planned to fully-automate
+all actions. For example, if `kubeone apply` determines that a node needs to be repaired,
+it would be up to the operator to create a new VM, update the cluster configuration
+(or Terraform state) and then run `kubeone apply` again.
+
+The following flowchart shows which actions are supported and how KubeOne makes the decision
+based on the results returned by probes.
+
+![kubeone apply flowchart](./apply-flowchart.svg)
+
+The first step in the reconciliation process is to determine is the cluster provisioned
+and the number of healthy nodes.
+
+This is done by running the following checks:
+* are all required packages installed (`docker`, `kubelet`, `kubeadm`)
+* is the `/etc/kubernetes` directory present and non-empty
+* is `kubelet` running and `/healthz` is reporting `ok`
+* are the control plane pods running and healthy
+* is a node part of the `etcd` ring
+
+#### The Cluster Is Not Provisioned
+
+This case is triggered when the `/etc/kubernetes` directory is absent or empty. This usually
+happens when newly created VMs are provided or `kubeone reset` has been run. In this case,
+`kubeone apply` would just invoke the installation process (`kubeone install`).
+
+It is important to make a difference between this case and the case when there is no healthy
+node in the cluster (but the cluster has been provisioned), as different steps should be taken.
+
+#### No Healthy Node Is Available or Quorum Is Not Satisfied
+
+If KubeOne determines that the number of healthy nodes are zero or not enough to
+satisfy the [quorum][etcd-quorum], KubeOne will return an error.
+
+In such a case, the `etcd` data integrity could **not** be guaranteed. It is
+required that operator take the following steps:
+* run `kubeone reset` to unprovision the cluster
+* run `kubeone apply` to reprovision the cluster
+* restore the backup
+
+An automated way for recovering from such scenario is a non-goal for this proposal.
+
+#### Non-Healthy Nodes are Present, but Quorum Is Satisfied
+
+If KubeOne determines that there are non-healthy nodes in the cluster, but
+the [quorum][etcd-quorum] is satisfied, KubeOne will attempt to automatically
+repair the cluster.
```

Comment:
Is such situation even possible? If a Node doesn't feel good, the etcd hosted on this Node will also be ether unreachable or ever worse — half-dead.